keras/src/optimizers/loss_scale_optimizer.py

```diff
@@ -171,8 +174,12 @@ def apply(self, grads, trainable_variables=None):
     def _stateful_handle_finite_grads(self, grads, trainable_variables):
         scale = self.dynamic_scale
         # Unscale gradients.
+        tvs = trainable_variables or self._trainable_variables
```

Comment:
You shouldn't need this. Line 185 it uses `trainable_variables` only.

Reply:
Can't it be `None` though?  It's called on line 211 with a default of `None`, which can be called from line 172, which also uses a default of `None`.

We need the actual `trainable_variables` to check the `overwrite_with_gradient` property.

Reply:
Ah yes, the line numbers changed when I uploaded a new commit.
```
apply(self, grads, trainable_variables=None)
```
calls
```
_common_apply(self, grads, trainable_variables=None)
```
which calls this function,
```
_stateful_handle_grads(self, grads, trainable_variables)
```
So if the `apply(...)` method is called with `trainable_variables=None` (or left absent to use the default `None`, then we end up in `_stateful_handle_grads(self, grads, None)`.  This is why I need to set:
```
tv = trainable_variables or self._trainable_variables.
```
if `trainable_variables` is `None` (because no trainable variables were passed to `apply`), then the variables are `self.trainable_variables`.

Reply:
Ok, I did not realize that it is optional in `apply`.